p-ata/src/processor.rs

```diff
@@ -0,0 +1,595 @@
+        cpi,
+        instruction::{AccountMeta, Instruction},
+        msg,
+        program_error::ProgramError,
+        pubkey::{find_program_address, Pubkey},
+        sysvars::{rent::Rent, Sysvar},
+        ProgramResult,
+    },
+    pinocchio_pubkey::derive_address,
+    spl_token_interface::state::{
+        account::Account as TokenAccount, mint::Mint, multisig::Multisig, Transmutable,
+    },
+};
+
+#[cfg(target_os = "solana")]
+use pinocchio::syscalls::sol_curve_validate_point;
+
+pub const INITIALIZE_ACCOUNT_3_DISCM: u8 = 18;
+pub const INITIALIZE_IMMUTABLE_OWNER_DISCM: u8 = 22;
+pub const TRANSFER_CHECKED_DISCM: u8 = 12;
+
+// Token-2022 AccountType::Account discriminator value
+const ACCOUNTTYPE_ACCOUNT: u8 = 2;
+
+pub const INITIALIZE_IMMUTABLE_OWNER_DATA: [u8; 1] = [INITIALIZE_IMMUTABLE_OWNER_DISCM];
+
+// Compile-time verifications
+const _: () = assert!(
+    TokenAccount::LEN == 165,
+    "TokenAccount size changed unexpectedly"
+);
+const _: () = assert!(Multisig::LEN == 355, "Multisig size changed unexpectedly");
+
+/// Parsed ATA accounts for create operations
+pub struct CreateAccounts<'a> {
+    pub payer: &'a AccountInfo,
+    pub associated_token_account_to_create: &'a AccountInfo,
+    pub wallet: &'a AccountInfo,
+    pub mint: &'a AccountInfo,
+    #[allow(dead_code)]
+    pub system_program: &'a AccountInfo,
+    pub token_program: &'a AccountInfo,
+    pub rent_sysvar: Option<&'a AccountInfo>,
+}
+
+/// Derive canonical ATA PDA from wallet, token program, and mint.
+///
+/// This is the least efficient derivation method, as it searches from bump
+/// 255 downward until an off-curve address is found. Use only when no bump hint
+/// is available in the instruction data.
+///
+/// An alternative was considered that used a loop with `derive_address` +
+/// `is_off_curve` instead of `find_program_address`, but though it saved ~30 CUs
+/// (1%) when bump happened to be `255`, it added more CUs on average.
+///
+/// ## Returns
+///
+/// `(address, bump)` - The canonical PDA address and its bump seed
+#[inline(always)]
+pub(crate) fn derive_canonical_ata_pda(
+    wallet: &Pubkey,
+    token_program: &Pubkey,
+    mint: &Pubkey,
+    program_id: &Pubkey,
+) -> (Pubkey, u8) {
+    find_program_address(
+        &[wallet.as_ref(), token_program.as_ref(), mint.as_ref()],
+        program_id,
+    )
+}
+
+/// Check if the given program ID is SPL Token (not Token-2022)
+#[inline(always)]
+pub(crate) fn is_spl_token_program(program_id: &Pubkey) -> bool {
+    const SPL_TOKEN_PROGRAM_ID: Pubkey =
+        pinocchio_pubkey::pubkey!("TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA");
+    *program_id == SPL_TOKEN_PROGRAM_ID
+}
+
+/// Check if account data represents an initialized token account.
+/// Mimics p-token's is_initialized_account check.
+///
+/// Safety: caller must ensure account_data.len() >= 109.
+#[inline(always)]
+pub(crate) unsafe fn is_initialized_account(account_data: &[u8]) -> bool {
```

Comment:
I don't think this is unsafe -- the slice will do the bounds check and panic appropriately. Did you mean to use `get_unchecked`?